File: python/divergent-modifications/good/financial_service.py
from datetime import datetime

class FinancialService:
    def __init__(self, db):
        self.calculator = FinancialCalculator()
        self.transaction_repository = TransactionRepository(db)
        self.user_repository = UserRepository(db)
        self.report_generator = ReportGenerator(self.transaction_repository, self.calculator)
        self.email_service = EmailService()

    # Financial calculations - delegated to FinancialCalculator
    def calculate_interest(self, principal: float, rate: float, time: int) -> float:
        return self.calculator.calculate_interest(principal, rate, time)

    def calculate_tax(self, income: float, deductions: float) -> float:
        return self.calculator.calculate_tax(income, deductions)

    # Database operations - delegated to repositories
    def save_transaction(self, user_id: int, amount: float, transaction_type: str) -> int:
        return self.transaction_repository.save_transaction(user_id, amount, transaction_type)

    def get_user_balance(self, user_id: int) -> float:
        return self.transaction_repository.get_user_balance(user_id)

    def update_user_profile(self, user_id: int, name: str, email: str) -> None:
        self.user_repository.update_user_profile(user_id, name, email)

    # Reporting - delegated to ReportGenerator
    def generate_monthly_report(self, user_id: int, month: int, year: int):
        return self.report_generator.generate_monthly_report(user_id, month, year)

    def generate_tax_report(self, user_id: int, year: int):
        return self.report_generator.generate_tax_report(user_id, year)

    # Email notifications - delegated to EmailService
    def send_monthly_statement(self, user_id: int) -> None:
        email = self.user_repository.get_user_email(user_id)
        if not email:
            return

        balance = self.get_user_balance(user_id)
        current_month = datetime.now().month
        current_year = datetime.now().year
        report = self.generate_monthly_report(user_id, current_month, current_year)

        self.email_service.send_monthly_statement(email, balance, report)

File: golang/divergent-modifications/good/financial_service.go
package main

import (
	"fmt"
	"time"
)

type FinancialCalculator interface {
	CalculateInterest(principal, rate float64, time int) float64
	CalculateTax(income, deductions float64) float64
}

type TransactionRepository interface {
	SaveTransaction(userId int, amount float64, transactionType string) (int64, error)
	GetUserBalance(userId int) float64
	GetMonthlyTransactions(userId, month, year int) []map[string]interface{}
	GetYearlyIncome(userId, year int) float64
	GetYearlyDeductions(userId, year int) float64
}

type UserRepository interface {
	UpdateUserProfile(userId int, name, email string) error
	GetUserEmail(userId int) (string, error)
}

type ReportGenerator interface {
	GenerateMonthlyReport(userId, month, year int) []map[string]interface{}
	GenerateTaxReport(userId, year int) map[string]interface{}
}

type EmailService interface {
	SendMonthlyStatement(email string, balance float64, monthlyReport []map[string]interface{})
}

type FinancialService struct {
	calculator           FinancialCalculator
	transactionRepository TransactionRepository
	userRepository       UserRepository
	reportGenerator      ReportGenerator
	emailService         EmailService
}

func NewFinancialService(
	calculator FinancialCalculator,
	transactionRepo TransactionRepository,
	userRepo UserRepository,
	reportGen ReportGenerator,
	emailSvc EmailService,
) *FinancialService {
	return &FinancialService{
		calculator:            calculator,
		transactionRepository: transactionRepo,
		userRepository:       userRepo,
		reportGenerator:      reportGen,
		emailService:         emailSvc,
	}
}

// Financial calculations - delegated to FinancialCalculator
func (fs FinancialService) CalculateInterest(principal, rate float64, time int) float64 {
	return fs.calculator.CalculateInterest(principal, rate, time)
}

func (fs FinancialService) CalculateTax(income, deductions float64) float64 {
	return fs.calculator.CalculateTax(income, deductions)
}

// Database operations - delegated to repositories
func (fs FinancialService) SaveTransaction(userId int, amount float64, transactionType string) (int64, error) {
	return fs.transactionRepository.SaveTransaction(userId, amount, transactionType)
}

func (fs FinancialService) GetUserBalance(userId int) float64 {
	return fs.transactionRepository.GetUserBalance(userId)
}

func (fs FinancialService) UpdateUserProfile(userId int, name, email string) error {
	return fs.userRepository.UpdateUserProfile(userId, name, email)
}

// Reporting - delegated to ReportGenerator
func (fs FinancialService) GenerateMonthlyReport(userId, month, year int) []map[string]interface{} {
	return fs.reportGenerator.GenerateMonthlyReport(userId, month, year)
}

func (fs FinancialService) GenerateTaxReport(userId, year int) map[string]interface{} {
	return fs.reportGenerator.GenerateTaxReport(userId, year)
}

// Email notifications - delegated to EmailService
func (fs FinancialService) SendMonthlyStatement(userId int) error {
	email, err := fs.userRepository.GetUserEmail(userId)
	if err != nil {
		return err
	}

	balance := fs.GetUserBalance(userId)
	currentTime := time.Now()
	report := fs.GenerateMonthlyReport(userId, int(currentTime.Month()), currentTime.Year())

	fs.emailService.SendMonthlyStatement(email, balance, report)
	return nil
}
